pennylane/shadows/classical_shadow.py

```diff
@@ -0,0 +1,370 @@
+    For example, we can compute the expectation value of a Pauli string
+
+    >>> shadow.expval(qml.PauliX(0) @ qml.PauliX(1), k=1)
+    (1.0079999999999998+0j)
+
+    or of a Hamiltonian:
+
+    >>> H = qml.Hamiltonian([1., 1.], [qml.PauliZ(0)@qml.PauliZ(1), qml.PauliX(0)@qml.PauliX(1)])
+    >>> shadow.expval(H, k=1)
+    (2.2319999999999998+0j)
+
+    The parameter ``k`` is used to estimate the expectation values via the `median of means` algorithm (see `2002.08953 <https://arxiv.org/abs/2002.08953>`_). The case ``k=1`` corresponds to simply taking the mean
+    value over all local snapshots. ``k>1`` corresponds to splitting the ``T`` local snapshots into ``k`` equal parts, and taking the median of their individual means.
+    """
+
+    def __init__(self, bits, recipes):
+        self.bits = bits
+        self.recipes = recipes
+
+        assert bits.shape == recipes.shape
+        self.snapshots = len(bits)
+
+        self.observables = [
+            qml.matrix(qml.PauliX(0)),
+            qml.matrix(qml.PauliY(0)),
+            qml.matrix(qml.PauliZ(0)),
+        ]
+
+    def local_snapshots(self, wires=None, snapshots=None):
+        r"""Compute the T x n x 2 x 2 local snapshots
+
+        For each qubit and each snapshot, compute :math:`3 U_i^\dagger |b_i \rangle \langle b_i| U_i - 1`
+
+        Args:
+            wires (Iterable[int]): The wires over which to compute the snapshots. For ``wires=None`` (default) all ``n`` qubits are used.
+            snapshots (Iterable[int] or int): Only compute a subset of local snapshots. For ``snapshots=None`` (default), all local snapshots are taken.
+                In case of an integer, a random subset of that size is taken. The subset can also be explicitly fixed by passing an Iterable with the corresponding indices.
+
+        Returns:
+            tensor: The local snapshots tensor of shape ``(T, n, 2, 2)`` containing the local local density matrices for each snapshot and each qubit.
+        """
+        bits = self.bits
+        recipes = self.recipes
+
+        if snapshots is not None:
+            if isinstance(snapshots, int):
+                # choose the given number of random snapshots
+                pick_snapshots = np.random.choice(
+                    np.arange(snapshots, dtype=np.int64), size=snapshots, replace=False
+                )
+            else:
+                # snapshots is an iterable that determines the indices
+                pick_snapshots = snapshots
+
+            pick_snapshots = qml.math.convert_like(pick_snapshots, bits)
+            bits = qml.math.gather(bits, pick_snapshots)
+            recipes = qml.math.gather(recipes, pick_snapshots)
+
+        if isinstance(wires, Iterable):
+            bits = bits[:, wires]
+            recipes = recipes[:, wires]
+
+        T, n = bits.shape
+
+        U = np.empty((T, n, 2, 2), dtype="complex")
+        for i, u in enumerate(self.observables):
+            U[np.where(recipes == i)] = u
```

Comment:
I believe the following has less time complexity:
```suggestion
        U = np.array([[self.observables[idx] for idx in recipe] for recipe in recipes], dtype="complex")
```

Reply:
Running locally, this actually seems a lot slower. I think there's some NumPy vectorization happening under the hood in the previous implementation.